incompatible with black v22.1.0 ?

```
$ poetry run darker .
Traceback (most recent call last):
  File "/home/jens/repos/dev-shell/.venv/bin/darker", line 8, in <module>
    sys.exit(main_with_error_handling())
  File "/home/jens/repos/dev-shell/.venv/lib/python3.9/site-packages/darker/__main__.py", line 421, in main_with_error_handling
    return main()
  File "/home/jens/repos/dev-shell/.venv/lib/python3.9/site-packages/darker/__main__.py", line 312, in main
    args, config, config_nondefault = parse_command_line(argv)
  File "/home/jens/repos/dev-shell/.venv/lib/python3.9/site-packages/darker/command_line.py", line 107, in parse_command_line
    config = load_config(args.src)
  File "/home/jens/repos/dev-shell/.venv/lib/python3.9/site-packages/darker/config.py", line 107, in load_config
    path = find_project_root(tuple(srcs or ["."])) / "pyproject.toml"
TypeError: unsupported operand type(s) for /: 'tuple' and 'str'
```

black changed the return value of find_project_root() here: https://github.com/psf/black/commit/521d1b8129c2d83b4ab49270fe7473802259c2a2#diff-1a5882c84fe4b3f18199a3e021822c1ea4bdc3f86952ea450c4db15d9387c7c3

It's not a tuple :(

Can comfirm with my project action
```
Run akaihola/darker@github-action-v1.3.2-2
Run # Exists since using github.action_path + path to main script doesn't work
Traceback (most recent call last):
  File "/home/runner/work/_actions/akaihola/darker/github-action-v1.3.2-2/.darker-env/bin/darker", line 8, in <module>
    sys.exit(main_with_error_handling())
  File "/home/runner/work/_actions/akaihola/darker/github-action-v1.3.2-2/.darker-env/lib/python3.8/site-packages/darker/__main__.py", line 421, in main_with_error_handling
    return main()
  File "/home/runner/work/_actions/akaihola/darker/github-action-v1.3.2-2/.darker-env/lib/python3.8/site-packages/darker/__main__.py", line 312, in main
    args, config, config_nondefault = parse_command_line(argv)
  File "/home/runner/work/_actions/akaihola/darker/github-action-v1.3.2-2/.darker-env/lib/python3.8/site-packages/darker/command_line.py", line 107, in parse_command_line
    config = load_config(args.src)
  File "/home/runner/work/_actions/akaihola/darker/github-action-v1.3.2-2/.darker-env/lib/python3.8/site-packages/darker/config.py", line 107, in load_config
    path = find_project_root(tuple(srcs or ["."])) / "pyproject.toml"
TypeError: unsupported operand type(s) for /: 'tuple' and 'str'
Error: Process completed with exit code 1.
```

--- a/.github/workflows/python-package.yml
+++ b/.github/workflows/python-package.yml
@@ -18,6 +18,11 @@ jobs:
       matrix:
         os: [ubuntu-latest, windows-latest, macos-latest]
         python-version: ['3.6', '3.7', '3.8', '3.9', '3.10']
+        constraints: ['']
+        include:
+        - os: ubuntu-latest
+          python-version: '3.7'
+          constraints: '--constraint constraints-oldest.txt'
 
     steps:
     - uses: actions/checkout@v2
@@ -30,7 +35,7 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade 'pip>=20.3'  # strict dependency resolution added in pip 20.3
-        pip install -e '.[isort,test]'
+        pip install -e '.[isort,test]' ${{ matrix.constraints }}
     - name: Test with pytest
       run: |
         pytest

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -12,6 +12,7 @@ Fixed
 - ``regex`` module now always available for unit tests
 - Compatibility with NixOS. Keep ``$PATH`` intact so Git can be called.
 - Updated tests to pass on new Pygments versions
+- Compatibility with Black 22.1
 
 
 1.3.2_ - 2021-10-28

--- a/constraints-oldest.txt
+++ b/constraints-oldest.txt
@@ -1,0 +1,13 @@
+# Constraints for pip to pin dependencies to oldest supported versions.
+# This is used in a GitHub Workflow matrix job which ensures everything
+# still works against oldest supported versions of both the Python
+# interpreter and Python ependencies. Keep this up-to-date with minimum
+# versions in `setup.cfg`.
+black==21.8b0
+mypy==0.910
+pytest==6.1.0
+pytest-flake8==1.0.6
+pytest-isort==1.1.0
+pytest-kwparametrize==0.0.3
+regex==2021.4.4
+types-toml==0.1.3

--- a/mypy.ini
+++ b/mypy.ini
@@ -31,6 +31,9 @@ strict_equality = True
 [mypy-darker.argparse_helpers]
 disallow_any_explicit = False
 
+[mypy-darker.black_compat]
+disallow_any_explicit = False
+
 [mypy-darker.command_line]
 disallow_any_explicit = False
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -26,10 +26,13 @@ package_dir =
     =src
 packages = find:
 install_requires =
+    # NOTE: remember to keep `constraints-oldest.txt` in sync with these
     black>=21.5b1
     toml
     typing-extensions ; python_version < "3.8"
     dataclasses ; python_version < "3.7"
+# NOTE: remember to keep `.github/workflows/python-package.yml` in sync
+#       with the minimum required Python version
 python_requires = >=3.6
 
 [options.packages.find]
@@ -46,6 +49,7 @@ pygments.lexers =
 isort =
     isort>=5.0.1
 test =
+    # NOTE: remember to keep `constraints-oldest.txt` in sync with these
     black>=21.7b1  # to prevent Mypy error about `gen_python_files`, see issue #189
     flake8<4
     mypy>=0.910

--- a/src/darker/black_compat.py
+++ b/src/darker/black_compat.py
@@ -19,6 +19,3 @@ def find_project_root(srcs: Sequence[str]) -> Path:
         return cast(Tuple[Path], root)[0]
     # Black < 22
     return cast(Path, root)
-
-
-find_project_root.cache_clear = black_find_project_root.cache_clear

--- a/src/darker/black_compat.py
+++ b/src/darker/black_compat.py
@@ -1,0 +1,24 @@
+"""Functions for maintaining compatibility with multiple Black versions"""
+
+from pathlib import Path
+from typing import Any, Sequence, Tuple, cast
+
+from black import find_project_root as black_find_project_root
+
+
+def find_project_root(srcs: Sequence[str]) -> Path:
+    """Hide changed return value type in Black behind this wrapper
+
+    :param srcs: Files and directories to find the common root for
+    :return: Project root path
+
+    """
+    root = cast(Any, black_find_project_root(tuple(srcs or ["."])))
+    if isinstance(root, tuple):
+        # Black >= 22.1
+        return cast(Tuple[Path], root)[0]
+    # Black < 22
+    return cast(Path, root)
+
+
+find_project_root.cache_clear = black_find_project_root.cache_clear

--- a/src/darker/config.py
+++ b/src/darker/config.py
@@ -7,7 +7,8 @@ from pathlib import Path
 from typing import Iterable, List, cast
 
 import toml
-from black import find_project_root
+
+from darker.black_compat import find_project_root
 
 if sys.version_info >= (3, 8):
     from typing import TypedDict

--- a/src/darker/import_sorting.py
+++ b/src/darker/import_sorting.py
@@ -3,8 +3,7 @@ import sys
 from pathlib import Path
 from typing import Any, Optional
 
-from black import find_project_root
-
+from darker.black_compat import find_project_root
 from darker.exceptions import IncompatiblePackageError, MissingPackageError
 from darker.utils import TextDocument
 

--- a/src/darker/tests/conftest.py
+++ b/src/darker/tests/conftest.py
@@ -6,8 +6,8 @@ from subprocess import check_call
 from typing import Dict, Optional
 
 import pytest
-from black import find_project_root
 
+from darker.black_compat import find_project_root
 from darker.git import _git_check_output_lines
 
 

--- a/src/darker/tests/conftest.py
+++ b/src/darker/tests/conftest.py
@@ -6,8 +6,8 @@ from subprocess import check_call
 from typing import Dict, Optional
 
 import pytest
+from black import find_project_root as black_find_project_root
 
-from darker.black_compat import find_project_root
 from darker.git import _git_check_output_lines
 
 
@@ -97,5 +97,11 @@ def git_repo(tmp_path, monkeypatch):
 
 @pytest.fixture
 def find_project_root_cache_clear():
-    """Clear LRU caching in :func:`black.find_project_root` before each test"""
-    find_project_root.cache_clear()
+    """Clear LRU caching in :func:`black.find_project_root` before each test
+
+    NOTE: We use `darker.black_compat.find_project_root` to wrap Black's original
+    function since its signature has changed along the way. However, clearing the cache
+    needs to be done on the original of course.
+
+    """
+    black_find_project_root.cache_clear()

--- a/src/darker/tests/test_black_diff.py
+++ b/src/darker/tests/test_black_diff.py
@@ -1,6 +1,7 @@
 """Unit tests for `darker.black_diff`"""
 
 import re
+from dataclasses import dataclass, field
 from pathlib import Path
 from unittest.mock import ANY, patch
 
@@ -15,6 +16,20 @@ from darker.black_diff import (
     run_black,
 )
 from darker.utils import TextDocument
+
+
+@dataclass
+class RegexEquality:
+    """Compare equality to either `re.Pattern` or `regex.Pattern`"""
+
+    pattern: str
+    flags: int = field(default=re.UNICODE)
+
+    def __eq__(self, other):
+        return (
+            other.pattern == self.pattern
+            and other.flags & 0x1FF == re.compile(self.pattern).flags | self.flags
+        )
 
 
 @pytest.mark.kwparametrize(
@@ -46,15 +61,19 @@ from darker.utils import TextDocument
     dict(config_lines=[r"include = '\.pyi$'"], expect={}),
     dict(
         config_lines=[r"exclude = '\.pyx$'"],
-        expect={"exclude": re.compile("\\.pyx$")},
+        expect={"exclude": RegexEquality("\\.pyx$")},
     ),
     dict(
         config_lines=["extend-exclude = '''", r"^/setup\.py", r"|^/dummy\.py", "'''"],
-        expect={"extend_exclude": re.compile("(?x)^/setup\\.py\n|^/dummy\\.py\n")},
+        expect={
+            "extend_exclude": RegexEquality(
+                "(?x)^/setup\\.py\n|^/dummy\\.py\n", re.VERBOSE
+            )
+        },
     ),
     dict(
         config_lines=["force-exclude = '''", r"^/setup\.py", r"|\.pyc$", "'''"],
-        expect={"force_exclude": re.compile("(?x)^/setup\\.py\n|\\.pyc$\n")},
+        expect={"force_exclude": RegexEquality("(?x)^/setup\\.py\n|\\.pyc$\n")},
     ),
     config_path=None,
 )

--- a/src/darker/tests/test_import_sorting.py
+++ b/src/darker/tests/test_import_sorting.py
@@ -7,7 +7,6 @@ from textwrap import dedent
 import pytest
 
 import darker.import_sorting
-from darker.black_compat import find_project_root
 from darker.tests.helpers import isort_present
 from darker.utils import TextDocument
 
@@ -76,9 +75,15 @@ def test_apply_isort(encoding, newline):
         ),
     ),
 )
-def test_isort_config(monkeypatch, tmpdir, line_length, settings_file, expect):
+def test_isort_config(
+    monkeypatch,
+    tmpdir,
+    find_project_root_cache_clear,
+    line_length,
+    settings_file,
+    expect,
+):
     """``apply_isort()`` parses ``pyproject.toml``correctly"""
-    find_project_root.cache_clear()
     monkeypatch.chdir(tmpdir)
     (tmpdir / "pyproject.toml").write(
         dedent(

--- a/src/darker/tests/test_import_sorting.py
+++ b/src/darker/tests/test_import_sorting.py
@@ -5,9 +5,9 @@ from pathlib import Path
 from textwrap import dedent
 
 import pytest
-from black import find_project_root
 
 import darker.import_sorting
+from darker.black_compat import find_project_root
 from darker.tests.helpers import isort_present
 from darker.utils import TextDocument
 

--- a/src/darker/tests/test_import_sorting.py
+++ b/src/darker/tests/test_import_sorting.py
@@ -1,5 +1,7 @@
 """Tests for :mod:`darker.import_sorting`"""
 
+# pylint: disable=unused-argument
+
 from importlib import reload
 from pathlib import Path
 from textwrap import dedent

--- a/src/darker/tests/test_main.py
+++ b/src/darker/tests/test_main.py
@@ -38,6 +38,12 @@ def test_isort_option_without_isort(git_repo, caplog):
 
 @pytest.fixture
 def run_isort(git_repo, monkeypatch, caplog, request, find_project_root_cache_clear):
+    """Fixture for running Darker with requested arguments and a patched `isort`
+
+    Provides an `run_isort.isort_code` mock object which allows checking whether and how
+    the `isort.code()` function was called.
+
+    """
     monkeypatch.chdir(git_repo.root)
     paths = git_repo.add({"test1.py": "original"}, commit="Initial commit")
     paths["test1.py"].write_bytes(b"changed")

--- a/src/darker/tests/test_main.py
+++ b/src/darker/tests/test_main.py
@@ -15,7 +15,6 @@ import pytest
 import darker.__main__
 import darker.import_sorting
 import darker.linting
-from darker.black_compat import find_project_root
 from darker.exceptions import MissingPackageError
 from darker.git import WORKTREE, RevisionRange
 from darker.tests.helpers import isort_present
@@ -38,9 +37,7 @@ def test_isort_option_without_isort(git_repo, caplog):
 
 
 @pytest.fixture
-def run_isort(git_repo, monkeypatch, caplog, request):
-    find_project_root.cache_clear()
-
+def run_isort(git_repo, monkeypatch, caplog, request, find_project_root_cache_clear):
     monkeypatch.chdir(git_repo.root)
     paths = git_repo.add({"test1.py": "original"}, commit="Initial commit")
     paths["test1.py"].write_bytes(b"changed")

--- a/src/darker/tests/test_main.py
+++ b/src/darker/tests/test_main.py
@@ -11,11 +11,11 @@ from types import SimpleNamespace
 from unittest.mock import call, patch
 
 import pytest
-from black import find_project_root
 
 import darker.__main__
 import darker.import_sorting
 import darker.linting
+from darker.black_compat import find_project_root
 from darker.exceptions import MissingPackageError
 from darker.git import WORKTREE, RevisionRange
 from darker.tests.helpers import isort_present

